### PR TITLE
[wasm] Fix regressed file sizes for blazor

### PR DESCRIPTION
**Pull request by radical**
[wasm] build: Revert to older behavior for WasmNativeStrip

The earlier change was done in 26ae09784c16dfa6707cd73eca2aeb00a28741d0,
which changed to pass `-g` to the link step also. But that resulted in
increased native file sizes.

- [wasm] Add getter for __indirect_function_table
- [wasm] cleanup corresponding tests
- [wasm] Remove WasmNativeStrip from wasm.proj as it will have no effect

Issue: Issue: https://github.com/dotnet/perf-autofiling-issues/issues/20642

### CODE REVIEW

File: src/mono/wasm/runtime/types/internal.ts [hunk 1]
```diff
@@ -457,6 +457,7 @@ export declare interface EmscriptenModuleInternal {
     ready: Promise<unknown>;
     asm: any;
     getMemory(): WebAssembly.Memory;
+    getWasmIndirectFunctionTable: any;
```
Comment (kg) [inline src/mono/wasm/runtime/types/internal.ts:460]: missing the ()

File: src/mono/wasm/runtime/es6/dotnet.es6.pre.js [hunk 2]
```diff
@@ -4,3 +4,4 @@ _nativeModuleLoaded = true;
 Module["getMemory"] = function () { return wasmMemory; }
 createDotnetRuntime = Module = createDotnetRuntime(Module);
```
Comment (kg) [inline src/mono/wasm/runtime/es6/dotnet.es6.pre.js:5]: Please also define it before createDotnetRuntime so it's available during initialization

**Review by kg — APPROVED** [review]

File: src/mono/wasm/runtime/es6/dotnet.es6.pre.js (hunk 2 above)
Comment (radical) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:5]: Before *and* after, like `getMemory`?

Comment (pavelsavara) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:6]: Let's not do it, unless we know about specific scenarios when this could happen.

Comment (kg) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:6]: I'm fine with Before Only if that's what you prefer

Comment (pavelsavara) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:6]: thanks

**Comment by radical** [comment]
After the change to move `getWasmIndirectFunctionTable` to only be before the runtime is created, library tests are failing with:

```
fail: [out of order message from the browser]: http://127.0.0.1:36071/_framework/dotnet.runtime.js 2:12344 "MONO_WASM: onRuntimeInitializedAsync() failed" TypeError: t.getWasmIndirectFunctionTable is not a function
          at br (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:55253)
          at zr (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:60544)
          at http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:214955
fail: [out of order message from the browser]: http://127.0.0.1:36071/_framework/dotnet.js 2:917 "MONO_WASM: TypeError: t.getWasmIndirectFunctionTable is not a function\n    at br (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:55253)\n    at zr (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:60544)\n    at http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:214955"
```
[log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-92664-merge-cd2156072dc14fa6b4/Wasm.Browser.Config.Sample/1/console.71f45b7b.log?helixlogtype=result).

**Comment by kg** [comment]
> After the change to move `getWasmIndirectFunctionTable` to only be before the runtime is created, ~library~ all the tests are failing with:
> 
> ```
> fail: [out of order message from the browser]: http://127.0.0.1:36071/_framework/dotnet.runtime.js 2:12344 "MONO_WASM: onRuntimeInitializedAsync() failed" TypeError: t.getWasmIndirectFunctionTable is not a function
>           at br (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:55253)
>           at zr (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:60544)
>           at http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:214955
> fail: [out of order message from the browser]: http://127.0.0.1:36071/_framework/dotnet.js 2:917 "MONO_WASM: TypeError: t.getWasmIndirectFunctionTable is not a function\n    at br (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:55253)\n    at zr (http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:60544)\n    at http://127.0.0.1:36071/_framework/dotnet.runtime.js:3:214955"
> ```
> 
> [log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-92664-merge-cd2156072dc14fa6b4/Wasm.Browser.Config.Sample/1/console.71f45b7b.log?helixlogtype=result).

I guess I put it in both places because it was actually necessary, and I forgot :-)

**Comment by radical** [comment]
Don't merge yet. The failures are relevant, because 9.0 runtime pack is being used with 8.0 targets.

File: src/mono/wasm/runtime/es6/dotnet.es6.pre.js [hunk 3]
```diff
@@ -2,5 +2,7 @@ if (_nativeModuleLoaded) throw new Error("Native module already loaded");
 _nativeModuleLoaded = true;
 // see https://github.com/emscripten-core/emscripten/issues/19832
 Module["getMemory"] = function () { return wasmMemory; }
+Module["getWasmIndirectFunctionTable"] = function () { return wasmTable; }
 createDotnetRuntime = Module = createDotnetRuntime(Module);
 Module["getMemory"] = function () { return wasmMemory; }
+Module["getWasmIndirectFunctionTable"] = function () { return wasmTable; }
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/es6/dotnet.es6.pre.js:8]: I'm ok with having the `getMemory` and `getWasmIndirectFunctionTable` before `createDotnetRuntime` but I don't understand why it needs to also be after.

**Review by pavelsavara — CHANGES_REQUESTED** [review]

File: src/mono/wasm/runtime/es6/dotnet.es6.pre.js (hunk 3 above)
Comment (radical) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:8]: https://github.com/dotnet/runtime/pull/92664#issuecomment-1738139068

Comment (pavelsavara) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:8]: I don't understand that comment, could you please explain why it's necessary ?

**Comment by pavelsavara** [comment]
> I guess I put it in both places because it was actually necessary, and I forgot :-)

@kg could you please explain how the method could get lost ?

File: src/mono/wasm/runtime/es6/dotnet.es6.pre.js (hunk 3 above)
Comment (kg) [reply src/mono/wasm/runtime/es6/dotnet.es6.pre.js:8]: I still don't actually understand why it's broken without this, TBH

**Comment by kg** [comment]
> > I guess I put it in both places because it was actually necessary, and I forgot :-)
> 
> @kg could you please explain how the method could get lost ?

If the "Module" global is re-created by createDotnetRuntime and our extensions aren't copied over, they would be missing from the new global and things would break. You want them there before initialization so they're always available, and you want to reinstall them after create if they are now missing. Fully hypothetical defensive coding though.

**Comment by pavelsavara** [comment]
> If the "Module" global is re-created by createDotnetRuntime

A) we can fix the merge of Module object if it's broken.

B) we can use `passEmscriptenInternals` in `dotnet.es6.lib.js` instead of capuring it in `dotnet.es6.pre.js`. I would prefer this because it would be consistent with all the other things we capture.


Like this 
image

**Review by pavelsavara — APPROVED** [review]

**Comment by radical** [comment]
Blazor tests [failing](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-92664-merge-897bcb2bc8bb404697/Workloads-Wasm.Build.Tests.Blazor.MiscTests3/1/console.1bde39ea.log?helixlogtype=result) with:

```
        [error] MONO_WASM: onRuntimeInitializedAsync() failed Error: raw cwrap mono_jiterp_placeholder_trace not found
            at Ur (http://127.0.0.1:46037/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58272)
            at http://127.0.0.1:46037/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213378
            at http://127.0.0.1:46037/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213662
        [error] MONO_WASM: Error: raw cwrap mono_jiterp_placeholder_trace not found
            at Ur (http://127.0.0.1:46037/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58272)
            at http://127.0.0.1:46037/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213378
            at http://127.0.0.1:46037/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213662
```

I had to fix some merge conflicts. Did I mess something up?

**Comment by pavelsavara** [comment]
> I had to fix some merge conflicts. Did I mess something up?

I don't see anything wrong with .ts

Is it possible that part of the build is `WasmBuildNative` true but not consistently so ?

File: src/mono/wasm/runtime/jiterpreter-support.ts [hunk 4]
```diff
@@ -1549,10 +1549,10 @@ export function copyIntoScratchBuffer(src: NativePointer, size: number): NativeP
 
 export function getWasmFunctionTable(module?: any) {
     const theModule = (<any>Module || module);
-    mono_assert (theModule, "Module not available yet");
-    mono_assert (theModule["asm"], "Module['asm'] not available yet");
+    mono_assert(theModule, "Module not available yet");
+    mono_assert(theModule["asm"], "Module['asm'] not available yet");
```
Comment (kg) [inline src/mono/wasm/runtime/jiterpreter-support.ts:1553]: We can remove this asm assertion now

**Review by kg — APPROVED** [review]

File: src/mono/wasm/runtime/jiterpreter-support.ts (hunk 4 above)
Comment (radical) [inline src/mono/wasm/runtime/jiterpreter-support.ts:1553]:
```suggestion
```

**Comment by radical** [comment]
WBT were failing with:
```
        [error] MONO_WASM: onRuntimeInitializedAsync() failed Error: raw cwrap mono_jiterp_placeholder_trace not found
            at Ur (http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58272)
            at http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213378
            at http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213662
```

File: src/mono/wasm/runtime/jiterpreter-support.ts [hunk 5]
```diff
@@ -1549,10 +1549,8 @@ export function copyIntoScratchBuffer(src: NativePointer, size: number): NativeP
 
 export function getWasmFunctionTable(module?: any) {
     const theModule = (<any>Module || module);
```
Comment (radical) [inline src/mono/wasm/runtime/jiterpreter-support.ts:1551]: Remove ununsed local.
```suggestion
```

**Comment by radical** [comment]
> WBT were failing with:
> 
> ```
>         [error] MONO_WASM: onRuntimeInitializedAsync() failed Error: raw cwrap mono_jiterp_placeholder_trace not found
>             at Ur (http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58272)
>             at http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213378
>             at http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213662
> ```

@kg these are failing with config=Debug, and AOT or relinking.

**Comment by kg** [comment]
> > WBT were failing with:
> > ```
> >         [error] MONO_WASM: onRuntimeInitializedAsync() failed Error: raw cwrap mono_jiterp_placeholder_trace not found
> >             at Ur (http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58272)
> >             at http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213378
> >             at http://127.0.0.1:37089/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:213662
> > ```
> 
> @kg these are failing with config=Debug, and AOT or relinking.

This should normally only be hit in the threading + AOT configuration, I don't know why it would be happening there. Will look into it

**Comment by kg** [comment]
> mono_jiterp_placeholder_trace

```c
EMSCRIPTEN_KEEPALIVE int
mono_jiterp_placeholder_trace (void *_frame, void *pLocals, JiterpreterCallInfo *cinfo, const guint16 *ip)
```
Is inside an `#ifdef HOST_BROWSER` block, so it shouldn't be missing unless the build is misconfigured. It's in interp.c.

**Comment by radical** [comment]
I might have messed up the merge. Let's see if CI can build it.

**Review by lewing — COMMENTED** [review]
import ENVIRONMENT_IS_NODE

**Comment by pavelsavara** [comment]
need help ?

**Comment by radical** [comment]
> need help ?

Any suggestions on a fix? The alternative would be to disable the failing tests which are essentially for Debug+AOT, and Debug+relink. We could ignore `debug+aot` for now, but debug+relink is likely more common because we would relink if you had any native references too.

But this PR will help the sizes.

**Comment by pavelsavara** [comment]
> Any suggestions on a fix? 

Still this, right ?
```
Wasm.Build.Tests.Blazor.SimpleRunTests.BlazorPublishRunTest(config: "Debug", aot: True) [FAIL]

 [error] MONO_WASM: onRuntimeInitializedAsync() failed Error: raw cwrap mono_jiterp_placeholder_trace not found
            at Ir (http://127.0.0.1:35275/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58604)
            at http://127.0.0.1:35275/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:214838
            at http://127.0.0.1:35275/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:215122
```

**Comment by radical** [comment]
> > Any suggestions on a fix? 
> 
> 
> 
> Still this, right ?
> 
> ```
> 
> Wasm.Build.Tests.Blazor.SimpleRunTests.BlazorPublishRunTest(config: "Debug", aot: True) [FAIL]
> 
> 
> 
>  [error] MONO_WASM: onRuntimeInitializedAsync() failed Error: raw cwrap mono_jiterp_placeholder_trace not found
> 
>             at Ir (http://127.0.0.1:35275/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:58604)
> 
>             at http://127.0.0.1:35275/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:214838
> 
>             at http://127.0.0.1:35275/_framework/dotnet.runtime.9.0.0-ci.uxcbg0k114.js:3:215122
> 
> ```

Yes

**Comment by pavelsavara** [comment]
I have this untested theory:

I think this is not about AOT, but about `WasmBuildNative` being triggered with some optimization level which mangles symbols. 

It fails in `getRawCwrap` which uses un-mangled symbol names.
`mono_jiterp_placeholder_trace` is just coincidence, because it's the first thing requested.

We have this un-mangling logic in `createWasmImportStubsFrom` and `replace_linker_placeholders` for wasm imports. Maybe we need same thing for wasm exports ?

**Comment by pavelsavara** [comment]
Printing full list of `Module["asm"]` would help to debug it.

**Comment by radical** [comment]
https://github.com/dotnet/runtime/pull/95613 seems to have fixed the failures! Thanks @kg, @pavelsavara!

**Comment by radical** [comment]
Failures are known issues.

**Comment by radical** [comment]
@microsoft-github-policy-service rerun

**Comment by radical** [comment]
@dotnet-policy-service rerun

**Pull request merged**